fog/graph/projection.py

```diff
@@ -0,0 +1,187 @@
+# =============================================================================
+# Fog Graph Projections
+# =============================================================================
+#
+# Miscellaneous functions related to bipartite to monopartite projections.
+#
+import math
+import networkx as nx
+from collections import defaultdict, Counter
+
+from fog.metrics.cosine import sparse_dot_product
+from fog.metrics.utils import intersection_size
+
+MONOPARTITE_PROJECTION_METRICS = ('cosine', 'jaccard', 'overlap')
+EMPTY_COUPLE = (None, None)
+
+
+def compute_metric(metric, vector1, vector2, norm1, norm2):
+    if metric == 'cosine':
+        w = sparse_dot_product(vector1, vector2)
+
+        if w == 0:
+            return 0
+
+        return w / (norm1 * norm2)
+
+    w = intersection_size(vector1, vector2)
+
+    if w == 0:
+        return 0
+
+    if metric == 'jaccard':
+        return w / (norm1 + norm2 - w)
+
+    if metric == 'overlap':
+        return w / min(norm1, norm2)
+
+    raise NotImplementedError
+
+
+# TODO: use passjoin prefix filtering as optimization scheme
+def monopartite_projection(bipartite, project, part='bipartite', weight='weight',
+                           metric=None, threshold=None, use_index=False,
+                           bipartition_check=True):
+    """
+    Function computing a monopartite projection of the given bipartite graph.
+    This projection can be basic and create a weighted edge each time two nodes
+    in target partition share a common neighbor. Or it can be weighted and
+    filtered using a similarity metric such as Jaccard or cosine similarity,
+    for instance.
+
+    Args:
+        bipartite (nx.Graph): Target bipartite graph.
+        project (str): Name of the partition to project.
+        part (str, optional): Name of the partition attribute.
+            Defaults to "bipartite" wrt networkx conventions.
+        weight (str, optional): Name of the weight attribute.
+            Defaults to "weight" wrt networkx conventions.
+        metric (str, optional): Metric to use. If `None`, the basic projection
+            will be returned. Also accepts `jaccard`, `overlap` or `cosine`.
+            Defaults to basic projection.
+        threshold (float, optional): Optional similarity threshold under which
```

Comment:
under which edges won't be added **to the monopartite projection** ? Otherwise it could be a threshold selecting the edges that will be considered in the input graph.